src/python/pants/backend/python/macros/poetry_requirements_test.py

```diff
@@ -0,0 +1,414 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from textwrap import dedent
+from typing import Iterable
+
+import pytest
+from _pytest.logging import LogCaptureFixture
+from pkg_resources import Requirement
+
+from pants.backend.python.macros.poetry_requirements import (
+    PoetryRequirements,
+    get_max_caret,
+    get_max_tilde,
+    handle_dict_attr,
+    handle_str_attr,
+    parse_pyproject_toml,
+    parse_single_dependency,
+)
+from pants.backend.python.target_types import PythonRequirementLibrary, PythonRequirementsFile
+from pants.base.specs import AddressSpecs, DescendantAddresses, FilesystemSpecs, Specs
+from pants.engine.addresses import Address
+from pants.engine.internals.scheduler import ExecutionError
+from pants.engine.target import Targets
+from pants.testutil.rule_runner import QueryRule, RuleRunner
+
```

Comment:
Good tests! It'd be helpful to also test modifiers on semver, like `1.2.3.dev0` and `1.2.3rc1`. Test that on all those version-related tests, like `test_caret`.

I was also wondering about calver: https://calver.org. Black is a notable project to use calver, e.g. `21.6b0`. It looks like Poetry treats that the same as SemVer by default though, e.g. `poetry add black` results in `^21.6b0`. So I don't think there's anything your macro should do special (and we'd have no way to know anways in the first place if it was calver).